File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int left = 0, ans = 0;
        unordered_map<char, int> chars;
        for (int right = 0; right < s.size(); ++right){
            ++chars[s[right]];
            while (chars[s[right]] > 1 && left <= right){
                --chars[s[left++]];
            }
            ans = max(ans, right - left + 1);
        }
        return ans;
    }
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        chars, left, ans = defaultdict(int), 0, 0
        
        for right in range(len(s)):
            chars[s[right]] += 1
            
            while chars[s[right]] > 1 and left <= right:
                chars[s[left]] -= 1
                left += 1
                
            ans = max(ans, right - left + 1)
        
        return ans